File: occasum-core/src/main/java/pw/stamina/occasum/properties/primitives/clamp/DoubleClamp.java
package pw.stamina.occasum.properties.primitives.clamp;

public interface DoubleClamp {

    double clamp(double value);

    static DoubleClamp min(double min) {
        return new MinDoubleClamp(min);
    }

    static DoubleClamp max(double max) {
        return new MaxDoubleClamp(max);
    }

    static DoubleClamp range(double min, double max) {
        return new RangeDoubleClamp(min, max);
    }
    
    static DoubleClamp none() {
        return NoClamp.INSTANCE;
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/primitives/clamp/DoubleClamp.kt
package pw.stamina.occasum.properties.primitives.clamp

interface DoubleClamp {

    fun clamp(value: Double): Double

    companion object {

        fun min(min: Double): DoubleClamp = MinDoubleClamp(min)

        fun max(max: Double): DoubleClamp = MaxDoubleClamp(max)

        fun range(min: Double, max: Double): DoubleClamp = RangeDoubleClamp(min, max)

        fun none(): DoubleClamp = NoClamp
    }
}
